server/src/gnome/secret_exchange.rs

```diff
@@ -0,0 +1,174 @@
+// String.
+
+use std::collections::HashMap;
+
+use base64::prelude::*;
+use oo7::{crypto, Key};
+use zeroize::{Zeroize, Zeroizing};
+
+const SECRET: &str = "secret";
+const PUBLIC: &str = "public";
+const PRIVATE: &str = "private";
+const IV: &str = "iv";
+const PROTOCOL: &str = "[sx-aes-1]\n";
+const CIPHER_TEXT_LEN: usize = 16;
+
+#[derive(Debug, Zeroize)]
+pub struct SecretExchange {
+    private_key: Key,
+    public_key: Key,
+}
+
+impl SecretExchange {
+    // Creates the initial payload containing caller public_key
+    pub fn begin(&self) -> String {
+        let map = HashMap::from([(PUBLIC, self.public_key.as_ref())]);
+
+        encode(&map)
+    }
+
+    // Creates the shared secret: an AES key
+    pub fn create_shared_secret(&self, exchange: &str) -> Result<String, crypto::Error> {
+        let decoded = decode(exchange)
+            .expect("SecretExchange decode error: failed to decode exchange string");
+        let server_public_key = Key::new(
+            decoded
+                .get(PUBLIC)
+                .expect("SecretExchange decode error: PUBLIC parameter is empty")
+                .to_vec(),
+        );
+        // Above two calls should never fail during SecretExchange
+        let aes_key =
+            Key::generate_aes_key_for_secret_exchange(&self.private_key, &server_public_key)?;
+        let map = HashMap::from([(PRIVATE, aes_key.as_ref())]);
+
+        Ok(encode(&map))
+    }
+
+    pub fn new() -> Result<Self, crypto::Error> {
+        let private_key = Key::generate_private_key()?;
+        let public_key = Key::generate_public_key_for_secret_exchange(&private_key)?;
+
+        Ok(Self {
+            private_key,
+            public_key,
+        })
+    }
+}
+
+// Converts a HashMap into a payload String
+fn encode(map: &HashMap<&str, &[u8]>) -> String {
+    let mut exchange = map
+        .iter()
+        .map(|(key, value)| format!("{}={}", key, BASE64_STANDARD.encode(value)))
+        .collect::<Vec<_>>()
+        .join("\n");
+    exchange.insert_str(0, PROTOCOL); // Add PROTOCOL prefix
+
+    exchange
+}
+
+// Converts a payload String into a HashMap
+fn decode(exchange: &str) -> Option<HashMap<&str, Vec<u8>>> {
+    let (_, exchange) = exchange.split_once(PROTOCOL)?; // Remove PROTOCOL prefix
+    let pairs = exchange.split("\n").collect::<Vec<_>>();
+    let mut map: HashMap<&str, Vec<u8>> = HashMap::new();
+
+    for pair in pairs {
+        if pair.is_empty() {
+            // To avoid splitting an empty line (last new line)
+            break;
+        }
+        let (key, value) = pair.split_once("=")?;
+        let encoded = BASE64_STANDARD.decode(value).unwrap_or(vec![]);
+        if encoded.is_empty() {
+            return None;
+        }
+        map.insert(key, encoded);
+    }
+
+    Some(map)
+}
+
+// Retrieves the secret from final secret exchange string
+pub(crate) fn retrieve_secret(exchange: &str, aes_key: &str) -> Option<Zeroizing<Vec<u8>>> {
```

Comment:
This should probably return a oo7::Secret.

Reply:
Yeah, it should. Done.